test/e2e/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,126 @@
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/iancoleman/strcase"
+	ispnv1 "github.com/infinispan/infinispan-operator/api/v1"
+	"github.com/infinispan/infinispan-operator/api/v2alpha1"
+	"github.com/infinispan/infinispan-operator/controllers"
+	"github.com/infinispan/infinispan-operator/controllers/constants"
+	batchtest "github.com/infinispan/infinispan-operator/test/e2e/batch"
+	tutils "github.com/infinispan/infinispan-operator/test/e2e/utils"
+	coreos "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	testKube = tutils.NewTestKubernetes(os.Getenv("TESTING_CONTEXT"))
+	// startChannel          = os.Getenv("SUBSCRIPTION_CHANNEL_START")
+	subName               = constants.GetEnvWithDefault("SUBSCRIPTION_NAME", "infinispan-operator")
+	subNamespace          = constants.GetEnvWithDefault("SUBSCRIPTION_NAMESPACE", tutils.Namespace)
+	sourceChannel         = constants.GetEnvWithDefault("SUBSCRIPTION_CHANNEL_SOURCE", "2.1.x")
+	latestChannel         = constants.GetEnvWithDefault("SUBSCRIPTION_CHANNEL_LATEST", "2.2.x")
+	catalogSource         = constants.GetEnvWithDefault("SUBSCRIPTION_CATALOG_SOURCE", "test-catalog")
+	catalogSourcNamespace = constants.GetEnvWithDefault("SUBSCRIPTION_CATALOG_SOURCE_NAMESPACE", tutils.Namespace)
+	subPackage            = constants.GetEnvWithDefault("SUBSCRIPTION_PACKAGE", "infinispan")
+	subStartingCsv        = constants.GetEnvWithDefault("SUBSCRIPTION_STARTING_CSV", "infinispan-operator.v2.1.5")
+
+	latestImage = constants.GetEnvWithDefault("LATEST_OPERAND_IMAGE", constants.DefaultOperandImageOpenJDK)
+)
+
+func TestUpgrade(t *testing.T) {
```

Comment:
We should probably add 0 pod kill at the end of the test to assert that cluster will recover from failover after upgrade